app/aboot/fs_boot.c

```diff
@@ -0,0 +1,146 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+#include <debug.h>
+#include <target.h>
+#include <string.h>
+
+#include <lib/bio.h>
+#include <lib/fs.h>
+
+#include "fs_boot.h"
+
+static const char *bootable_parts[] = {
+	"system",
+	"cache",
+	"userdata",
+};
+static const int BOOTABLE_PART_COUNT = 3;
+
+static bool fsboot_bootable_part(char *label)
+{
+	int i;
+	for (i = 0; i < BOOTABLE_PART_COUNT; ++i)
+		if (strcmp(label, bootable_parts[i]) == 0)
+			return true;
+
+	return false;
+}
+
+/* Only iterate through files if target is NULL */
+static int fsboot_fs_load_img(char *dev_name, void* target, size_t sz)
+{
+	struct dirhandle *dirh;
+	struct dirent dirent;
+
+	char image_path[128] = "/mnt/";
+	int ret = 0, error = -2;
+
+	if (fs_mount("/mnt", "ext2", dev_name) >= 0) {
+		error = 0;
+		ret = fs_open_dir("/mnt", &dirh);
+		if (ret >= 0) {
+			while (fs_read_dir(dirh, &dirent) >= 0) {
+				if (!target) {
+					dprintf(SPEW, "| /%s/%s\n", dev_name, dirent.name);
+				} else if (strncmp(dirent.name, "boot.img", 7) == 0) {
+					strcpy(image_path, "/mnt/");
+					strcat(image_path, dirent.name);
```

Comment:
Although I guess you could just `strcpy(&image_path[strlen("/mnt/")], dirent.name)` here, hm.